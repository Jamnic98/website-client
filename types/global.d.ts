export type NavlinkData = {
  label: string;
  url: string;
  children: navlinkObj[];
};


export type RunData = {
  distance: number;
  duration: number;
  start_date_local: date;
};
