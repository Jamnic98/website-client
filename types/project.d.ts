export type ProjectLinkData = { URL: string; label: string; type: string };

export type ProjectData = {
  id: string;
  title: string;
  summary: string;
  paragraphs: string[];
  projectPageURI: string;
  screenshotURIs: string[];
  mainLanguage: string;
  links: ProjectLinkData[];
  techStack: string[];
};
