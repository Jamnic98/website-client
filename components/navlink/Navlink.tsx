import React, {type FC, useState} from 'react'
import Link from 'next/link'

import {Caret, Dropdown} from '..'
import {type NavlinkDataType} from '../../types'
import {useWidth} from '../../utils'
import styles from './Navlink.module.css'

interface NavlinkProps {
	navlinkObj: NavlinkDataType
	onClick: Function
}

export const Navlink: FC<NavlinkProps> = ({
	navlinkObj,
	onClick: handleClick,
}) => {
	const [isActive, setIsActive] = useState(false)
	const {label, url, children} = navlinkObj
	const screenWidth = useWidth() || 800
	return (
		<>
			{screenWidth >= 800 && children.length ? (
				<div
					className={styles.dropdownNavlink}
					onMouseOver={() => setIsActive(true)}
					onMouseOut={() => setIsActive(false)}
					onClick={() => handleClick()}
				>
					<Link className={styles.navlink} href={url}>
						{label}
						<Caret isFlipped={isActive} />
					</Link>
					<div
						className={styles.dropdownContainer}
						data-testid="dropdownContainer"
					>
						<Dropdown
							isOpen={isActive}
							items={children}
							render={(navlinkObj: NavlinkDataType) => (
								<Link href={navlinkObj.url} legacyBehavior>
									<a className={styles.navlink}>{navlinkObj.label}</a>
								</Link>
							)}
						/>
					</div>
				</div>
			) : (
				<Link href={url} legacyBehavior>
					<a onClick={() => handleClick()} className={styles.navlink}>
						{label}
					</a>
				</Link>
			)}
		</>
	)
}
