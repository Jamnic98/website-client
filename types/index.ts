export * from './navlink.types'
export * from './project.types'
export * from './runs.types'
