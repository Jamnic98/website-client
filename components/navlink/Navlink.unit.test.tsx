import React from 'react'
import renderer from 'react-test-renderer'
import {render, screen, fireEvent} from '@testing-library/react'

import {Navlink} from '../navlink'
import {NavlinkDataType} from '../../types/global'

const mockOnClick = jest.fn()

const testNavlinkObj: NavlinkDataType = {
	label: 'logs',
	url: '/logs',
	children: [{label: 'running', url: '/logs/running', children: []}],
}

describe('Navlink', () => {
	beforeAll(() => {
		global.innerWidth = 1200
		global.dispatchEvent(new Event('resize'))
	})

	it('renders Navlink', () => {
		const testNavlink = (
			<Navlink navlinkObj={testNavlinkObj} onClick={mockOnClick} />
		)

		render(testNavlink)
		const navlink = screen.getByText('logs')
		expect(navlink).toBeInTheDocument()
	})

	it('opens and closes dropdown onMouseOver and onMouseOut', () => {
		const testNavlink = (
			<Navlink navlinkObj={testNavlinkObj} onClick={mockOnClick} />
		)

		render(testNavlink)
		const navlink = screen.getByText('logs')
		const dropdown = screen.getByTestId('dropdown')
		const dropdownContainer = screen.getByTestId('dropdownContainer')

		expect(dropdown).toHaveClass('dropdown closed')
		fireEvent.mouseOver(navlink)
		expect(dropdown).toHaveClass('dropdown open')
		fireEvent.mouseOut(navlink)
		expect(dropdown).toHaveClass('dropdown closed')

		fireEvent.mouseOver(dropdownContainer)
		expect(dropdown).toHaveClass('dropdown open')
		fireEvent.mouseOut(dropdownContainer)
		expect(dropdown).toHaveClass('dropdown closed')
	})

	it('renders Navlink unchanged', () => {
		const tree = renderer
			.create(<Navlink navlinkObj={testNavlinkObj} onClick={mockOnClick} />)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})
})
