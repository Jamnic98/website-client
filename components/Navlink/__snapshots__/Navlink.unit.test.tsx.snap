// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navlink renders Navlink unchanged 1`] = `
<div
  className="dropdownNavlink"
>
  <a
    className="navlink"
    href="/logs"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    logs
    <span
      className="caret"
    >
      ^
    </span>
  </a>
  <div
    className="dropdownContainer"
    data-testid="dropdownContainer"
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="dropdown closed"
      data-testid="dropdown"
    >
      <div
        className="dropdownItem"
      >
        <a
          className="navlink"
          href="/logs/running"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          running
        </a>
      </div>
    </div>
  </div>
</div>
`;
